[JSC] Implement `Set.prototype.isSupersetOf` in C++

sosukesuzuki opened the pull request:
#### 54ed7bdc67272428e6c057e65aac9f7c8ac1fa3a
<pre>
[JSC] Implement `Set.prototype.isSupersetOf` in C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=298319">https://bugs.webkit.org/show_bug.cgi?id=298319</a>

Reviewed by Yusuke Suzuki.

This patch changes to implement `Set.prototype.isSupersetOf` in C++.

                          TipOfTree                  Patched

set-isSupersetOf        5.0179+-0.1692     ^      1.8690+-0.1636        ^ definitely 2.6848x faster

* JSTests/stress/set-prototype-isSupersetOf-empty-sets.js: Added.
(shouldBe):
* JSTests/stress/set-prototype-isSupersetOf-generic-object.js: Added.
(shouldBe):
(let.setLikeObject.has):
(let.setLikeObject.return.next):
(let.setLikeObject.keys):
(let.partialMatch.has):
(let.partialMatch.return.next):
(let.partialMatch.keys):
(let.tooBig.has):
(let.tooBig.return.next):
(let.tooBig.keys):
(shouldThrow.set1.isSupersetOf):
(shouldThrow):
* JSTests/stress/set-prototype-isSupersetOf-iterator-effects.js: Added.
(shouldBe):
(main.let.otherObject.has):
(main.let.otherObject.return.next):
(main.let.otherObject.keys):
(testIteratorModification.let.otherObject.has):
(testIteratorModification.let.otherObject.return.next):
(testIteratorModification.let.otherObject.keys):
(testIteratorModification):
* JSTests/stress/set-prototype-isSupersetOf-large-sets.js: Added.
(shouldBe):
(let.perfSetLike.has):
(let.perfSetLike.return.next):
(let.perfSetLike.keys):
(set let):
* JSTests/stress/set-prototype-isSupersetOf-property-access-order.js: Added.
(shouldBe):
(main.let.obj.get size):
(main.let.obj.get has):
(main.let.obj.get keys.return.next):
(main.let.obj.get keys):
(main.let.set new):
(testEarlyReturn.let.obj.get size):
(testEarlyReturn.let.obj.get has):
(testEarlyReturn.let.obj.get keys):
(testEarlyReturn):
* JSTests/stress/set-prototype-isSupersetOf-reference.js: Added.
(shouldBe):
(main.try.set isSupersetOf):
(main.try.set isSupersetOf.):
(main.const.set new):
(main.set catch):
(main):
* JSTests/stress/set-prototype-isSupersetOf-special-values.js: Added.
(shouldBe):
(let.setLike.has):
(let.setLike.return.next):
(let.setLike.keys):
(let.set new):
(let.symbolSetLike.has):
(let.symbolSetLike.return.next):
(let.symbolSetLike.keys):
(let.result.set isSupersetOf):
(let.objectSetLike.has):
(let.objectSetLike.return.next):
(let.objectSetLike.keys):
(let.objectSetLikeDifferent.has):
(let.objectSetLikeDifferent.return.next):
(let.objectSetLikeDifferent.keys):
* Source/JavaScriptCore/builtins/SetPrototype.js:
(isSupersetOf.wrapper.iterator): Deleted.
(isSupersetOf): Deleted.
* Source/JavaScriptCore/runtime/SetPrototype.cpp:
(JSC::SetPrototype::finishCreation):
(JSC::fastSetIsSupersetOf):
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/299667@main">https://commits.webkit.org/299667@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fe94e6147dbf1be11106cb82706ba663bf8da203

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119441 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39128 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29783 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125726 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71512 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39825 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47709 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90769 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60066 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122393 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31786 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107091 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71247 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30822 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25200 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69357 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/111564 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101243 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25391 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128686 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/117957 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46359 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35093 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99333 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46724 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103290 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99145 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25242 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44579 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22585 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42928 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46222 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51922 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/146657 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45687 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37703 "Found 1 new JSC binary failure: testapi, Found 18620 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_ctr.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_slice.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_sort2.js.default, ChakraCore.yaml/ChakraCore/test/Array/reverse1.js.default, ChakraCore.yaml/ChakraCore/test/Array/toLocaleString.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/blue_1086262.js.default, ChakraCore.yaml/ChakraCore/test/Date/date_cache_bug.js.default, ChakraCore.yaml/ChakraCore/test/Error/CallNonFunction.js.default, ChakraCore.yaml/ChakraCore/test/Error/stack2.js.default, ChakraCore.yaml/ChakraCore/test/Function/FuncBody.bug227901.js.default ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49037 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47374 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/SetPrototype.cpp:
```diff
@@ -558,6 +559,128 @@ JSC_DEFINE_HOST_FUNCTION(setProtoFuncIsSubsetOf, (JSGlobalObject* globalObject,
+        RETURN_IF_EXCEPTION(scope, { });
+        if (done.toBoolean(globalObject))
+            break;
+        RETURN_IF_EXCEPTION(scope, { });
```
This looks mot correct. If toBoolean does not throw, then we do not need it. If it is not, then this is required before this if statement (or using EXCEPTION_ASSERT)

sosukesuzuki replied:
Thanks, fixed!

sosukesuzuki committed:
[JSC] Implement `Set.prototype.isSupersetOf` in C++
https://bugs.webkit.org/show_bug.cgi?id=298319

Reviewed by Yusuke Suzuki.

This patch changes to implement `Set.prototype.isSupersetOf` in C++.

                          TipOfTree                  Patched

set-isSupersetOf        5.0179+-0.1692     ^      1.8690+-0.1636        ^ definitely 2.6848x faster

* JSTests/stress/set-prototype-isSupersetOf-empty-sets.js: Added.
(shouldBe):
* JSTests/stress/set-prototype-isSupersetOf-generic-object.js: Added.
(shouldBe):
(let.setLikeObject.has):
(let.setLikeObject.return.next):
(let.setLikeObject.keys):
(let.partialMatch.has):
(let.partialMatch.return.next):
(let.partialMatch.keys):
(let.tooBig.has):
(let.tooBig.return.next):
(let.tooBig.keys):
(shouldThrow.set1.isSupersetOf):
(shouldThrow):
* JSTests/stress/set-prototype-isSupersetOf-iterator-effects.js: Added.
(shouldBe):
(main.let.otherObject.has):
(main.let.otherObject.return.next):
(main.let.otherObject.keys):
(testIteratorModification.let.otherObject.has):
(testIteratorModification.let.otherObject.return.next):
(testIteratorModification.let.otherObject.keys):
(testIteratorModification):
* JSTests/stress/set-prototype-isSupersetOf-large-sets.js: Added.
(shouldBe):
(let.perfSetLike.has):
(let.perfSetLike.return.next):
(let.perfSetLike.keys):
(set let):
* JSTests/stress/set-prototype-isSupersetOf-property-access-order.js: Added.
(shouldBe):
(main.let.obj.get size):
(main.let.obj.get has):
(main.let.obj.get keys.return.next):
(main.let.obj.get keys):
(main.let.set new):
(testEarlyReturn.let.obj.get size):
(testEarlyReturn.let.obj.get has):
(testEarlyReturn.let.obj.get keys):
(testEarlyReturn):
* JSTests/stress/set-prototype-isSupersetOf-reference.js: Added.
(shouldBe):
(main.try.set isSupersetOf):
(main.try.set isSupersetOf.):
(main.const.set new):
(main.set catch):
(main):
* JSTests/stress/set-prototype-isSupersetOf-special-values.js: Added.
(shouldBe):
(let.setLike.has):
(let.setLike.return.next):
(let.setLike.keys):
(let.set new):
(let.symbolSetLike.has):
(let.symbolSetLike.return.next):
(let.symbolSetLike.keys):
(let.result.set isSupersetOf):
(let.objectSetLike.has):
(let.objectSetLike.return.next):
(let.objectSetLike.keys):
(let.objectSetLikeDifferent.has):
(let.objectSetLikeDifferent.return.next):
(let.objectSetLikeDifferent.keys):
* Source/JavaScriptCore/builtins/SetPrototype.js:
(isSupersetOf.wrapper.iterator): Deleted.
(isSupersetOf): Deleted.
* Source/JavaScriptCore/runtime/SetPrototype.cpp:
(JSC::SetPrototype::finishCreation):
(JSC::fastSetIsSupersetOf):
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/299667@main
